toolchain/check/sem_ir_diagnostic_converter.cpp

```diff
@@ -9,7 +9,21 @@ namespace Carbon::Check {
 
 auto SemIRDiagnosticConverter::ConvertLoc(SemIRLoc loc,
                                           ContextFnT context_fn) const
-    -> DiagnosticLoc {
+    -> std::pair<DiagnosticLoc, int32_t> {
```

Comment:
Have you considered adding the offset as a field in `DiagnosticLoc`? I'm not sure whether that'd be reasonable or not, so if this doesn't work out for some reason I think it's fine as-is.

Reply:
A given Diagnostic can have multiple DiagnosticMessages, each with their own DiagnosticLoc. The DiagnosticLoc will often point at a line in a different file as compared to the file the last_byte_offset points into. I believe putting the last_byte_offset on the Diagnostic avoids the potential for confusion about what a last_byte_offset on a note in a different file means.

Reply:
I was also wondering the same, since they are created and travel together so much. Might it help if the DiagnosticLoc field was an "ordering key" instead of an explicit "offset" which may be from a different file?

Reply:
In a diagnostic message such as:

```
  // CHECK:STDERR: fail_private_field_access.carbon:[[@LINE+7]]:21: error: cannot access private member `radius` of type `Circle` [ClassInvalidMemberAccess]
  // CHECK:STDERR:   let radius: i32 = circle.radius;
  // CHECK:STDERR:                     ^~~~~~~~~~~~~
```

The portions provided by DiagnosticLoc are everything that relates to location:

- filename (`fail_private_field_access.carbon`)
- line number (`[[@LINE+7]]`)
- column number (`21`)
- length (`^~~~~~~~~~~~~`)

I think I have a different perspective here. We have a few functions returning `DiagnosticLoc` + `last_byte_offset`, but the code with `DiagnosticLoc` only I am not touching.

Reply:
Note, if there's a strong preference, I can add a wrapper struct instead of using std::pair; but I do think DiagnosticLoc should remain separate

Reply:
Yeah that's fair. I also thought about SortedDiagonsticLoc in place of the pair, or whatever would make sense. A bare `int32_t` in a pair is fairly non-descriptive.

Reply:
Adding a wrapper struct sounds like a nice approach, as it'd give us a field name and a place to put documentation of the value separate from the `ConvertLoc` function.

Reply:
I'd done this last night, but got distracted -- pushed the change for it.